### PR TITLE
Dynamic declared class

### PR DESCRIPTION
**Pull request by thogg4**
#1610 causes the params to be passed to `declared` in the correct class, but once they get there that method only returns a `Hash`

### CODE REVIEW

**Comment by dblock** [comment]
Makes sense, but needs a test/changelog/etc., please.

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 #### Fixes
 
+* [#1631](https://github.com/ruby-grape/grape/pull/1631): Dynamic declared class - [@thogg4](https://github.com/thogg4).
```
Comment (dblock) [inline CHANGELOG.md]: Maybe something more useful?

File: spec/grape/endpoint_spec.rb [hunk 2]
```diff
@@ -298,6 +298,65 @@ def app
...
+
+          requires :first
+          optional :second, default: 'second-default'
+        end
+        subject.get '/declared' do
+          d = declared(params, include_missing: true)
+          ''
+        end
+
+        puts d
```
Comment (nbulaj) [inline spec/grape/endpoint_spec.rb]: I think there is no need in that piece of code, and the same for above

File: spec/grape/endpoint_spec.rb [hunk 3]
```diff
@@ -298,6 +298,65 @@ def app
...
+        d = nil
+
+        subject.params do
+          build_with Grape::Extensions::Hash::ParamBuilder
+
+          requires :first
+          optional :second, default: 'second-default'
+        end
+        subject.get '/declared' do
+          d = declared(params, include_missing: true)
```
Comment (dblock) [inline spec/grape/endpoint_spec.rb]: You should return the declared output and compare that, don't use a variable out of scope.

File: spec/grape/endpoint_spec.rb [hunk 4]
```diff
@@ -298,6 +298,65 @@ def app
       end
     end
 
+    context 'when params are not built with default class' do
+      it 'should return an object that corresponds with the params class - hash with indifferent access' do
```
Comment (dblock) [inline spec/grape/endpoint_spec.rb]: Instead of "should return", just say "returns"

File: spec/grape/endpoint_spec.rb (hunk 3 above)
Comment (thogg4) [inline spec/grape/endpoint_spec.rb]: @dblock what is returned by the api call is always a json string. What we are testing here is what is returned by `declared`. Is there a better place to put this to test just that?

Comment (dblock) [inline spec/grape/endpoint_spec.rb]: Have the API return `d.class` as JSON or a string and compare the result.

Comment (thogg4) [inline spec/grape/endpoint_spec.rb]: @dblock duh, im not sure why i didn't think of that. thanks 😄

File: spec/grape/endpoint_spec.rb [hunk 5]
```diff
@@ -298,6 +298,56 @@ def app
       end
     end
 
+    context 'when params are not built with default class' do
+      it 'returns an object that corresponds with the params class - hash with indifferent access' do
+        subject.params do
+          build_with Grape::Extensions::ActiveSupport::HashWithIndifferentAccess::ParamBuilder
+
+          requires :first
```
Comment (dblock) [inline spec/grape/endpoint_spec.rb]: The actual parameters are not material for the tests, remove them. Make this as short as possible.

**Comment by dblock** [comment]
Looks like there's a legit failure with one of the builds, check it out until this build is green please.

**Comment by dblock** [comment]
This looks good, just needs a rebase/squash please.

**Pull request closed without merging**